File: 653_find_target/find_target.go
/*
 * @Date: 2022-03-21 00:26:30
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-21 04:31:35
 * @FilePath: /algorithm/653_find_target/find_target.go
 */

package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findTarget(root *TreeNode, k int) bool {
	left, right := root, root
	leftStk := []*TreeNode{left}
	for left.Left != nil {
		leftStk = append(leftStk, left.Left)
		left = left.Left
	}
	rightStk := []*TreeNode{right}
	for right.Right != nil {
		rightStk = append(rightStk, right.Right)
		right = right.Right
	}
	for left != right {
		sum := left.Val + right.Val
		if sum == k {
			return true
		}
		if sum < k {
			left = leftStk[len(leftStk)-1]
			leftStk = leftStk[:len(leftStk)-1]
			for node := left.Right; node != nil; node = node.Left {
				leftStk = append(leftStk, node)
			}
		} else {
			right = rightStk[len(rightStk)-1]
			rightStk = rightStk[:len(rightStk)-1]
			for node := right.Left; node != nil; node = node.Right {
				rightStk = append(rightStk, node)
			}
		}
	}
	return false
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}

	root := &TreeNode{Val: 5, Left: &TreeNode{Val: 3, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 4}}, Right: &TreeNode{Val: 6, Right: &TreeNode{Val: 7}}}
	assert(findTarget(root, 9) == true)
	assert(findTarget(root, 28) == false)
}

File: cpp/653_find_target/find_target.cpp
/*
 * @Date: 2022-03-21 00:26:26
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-03-21 01:09:41
 * @FilePath: /algorithm/653_find_target/find_target.cpp
 */

#include <cassert>
#include <stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  TreeNode *getLeft(stack<TreeNode *> &stk) {
    TreeNode *root = stk.top();
    stk.pop();
    TreeNode *node = root->right;
    while (node != nullptr) {
      stk.push(node);
      node = node->left;
    }
    return root;
  }

  TreeNode *getRight(stack<TreeNode *> &stk) {
    TreeNode *root = stk.top();
    stk.pop();
    TreeNode *node = root->left;
    while (node != nullptr) {
      stk.push(node);
      node = node->right;
    }
    return root;
  }

  bool findTarget(TreeNode *root, int k) {
    TreeNode *left = root, *right = root;
    stack<TreeNode *> leftStack, rightStack;
    leftStack.push(left);
    while (left->left != nullptr) {
      leftStack.push(left->left);
      left = left->left;
    }
    rightStack.push(right);
    while (right->right != nullptr) {
      rightStack.push(right->right);
      right = right->right;
    }
    while (left != right) {
      if (left->val + right->val == k) {
        return true;
      }
      if (left->val + right->val < k) {
        left = getLeft(leftStack);
      } else {
        right = getRight(rightStack);
      }
    }
    return false;
  }
};

int main() {
  TreeNode *root =
      new TreeNode(5, new TreeNode(3, new TreeNode(2), new TreeNode(4)),
                   new TreeNode(6, nullptr, new TreeNode(7)));
  assert(Solution().findTarget(root, 9) == true);
  assert(Solution().findTarget(root, 28) == false);
}